### PR TITLE
updated the cli to work similar as the maven and gradle tools

### PR DESCRIPTION
**Pull request by stalep**
To test: ```java -jar target/quarkus-cli-999-SNAPSHOT.jar```
That should list the possible commands. Use ```--help``` for additional info for each command.
If you run eg: ```java -jar target/quarkus-cli-999-SNAPSHOT.jar create-project``` it will prompt the user for different values, not all the options are prompted atm.
To test interactive mode, use: ```java -jar target/quarkus-cli-999-SNAPSHOT.jar -i```

To create a native image use: ```mvn install -Dnative-image```

### CODE REVIEW

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java
Comment (aloubyansky) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java:43]: I guess, it's fine to keep as you are doing a refactoring here but this getGlobalDefault() should not be used. We should get rid of it ASAP. You may be the only one using it now.

Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java:43]: I see, what is the preferred way of doing it?
There are quite a few changes since I last looked at this so I just used what I thought worked :)

**Review by aloubyansky — APPROVED** [review]
I don't think there is a reason not to merge these changes but there is a bit more work to do. I think the add and list commands are not actually using the platform(s) from the project but whatever is returned by `getGlobalDefault()` which you may have assumed it took care of it :)

**Comment by stalep** [comment]
Yeah, as I mentioned there are a lot of changes I haven't seen before so I just used what I thought worked. There are also some options that's only defined atm, but not used yet (especially for dev-mode). - I just wanted to add an update first and some more incremental improvements going forward.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java
Comment (aloubyansky) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java:43]: We are missing a convenient abstraction for that, which is going to complicate it for you and any non-maven/gradle tool. For Maven projects we do that in BuildFileMojoBase (look how we initialize platformDescr). In Gradle that would be QuarkusPlatformTask.platformDescriptor().

**Comment by ia3andy** [comment]
Please wait before merging, I am about to provide a big refactor on those too :-/

**Comment by stalep** [comment]
> Please wait before merging, I am about to provide a big refactor on those too :-/
On what? :)

**Comment by ia3andy** [comment]
We will have a few conflicts, but not a lot, I guess it would be better to adapt this one than the other way around. I am fixing a last bug before providing my PR and we can decide the order then?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/AddExtensionCommand.java:43]: I see, yes it is a bit difficult on the CLI side since the project can be either Maven/Gradle. Some moons ago I used BuildFile/BuildTool to try to figure out what it was. If we could have a fairly simple way to do that it would be helpful for the CLI.

**Comment by ia3andy** [comment]
I need to review it first, but I published it as a draft so you can get an idea of it @stalep: https://github.com/quarkusio/quarkus/pull/9787

**Comment by stalep** [comment]
I see that there will be a fair bit of conflicts yes :)
I don't mind having your PR merged first and I can rebase this.

**Comment by ia3andy** [comment]
> I see that there will be a fair bit of conflicts yes :)
> I don't mind having your PR merged first and I can rebase this.

Thanks 😅

**Comment by maxandersen** [comment]
Thanks @stalep will give it a try :)

File: devtools/aesh/pom.xml
Comment (maxandersen) [inline devtools/aesh/pom.xml:23]: will it ever be used ?

File: devtools/aesh/reflectconfigs/quarkuscli.json
Comment (maxandersen) [inline devtools/aesh/reflectconfigs/quarkuscli.json:6]: someone should make a quarkus aesh extension making this not needed :)

File: devtools/aesh/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/QuarkusCli.java:9]: any reason this is not a quarkus command mode main ?

Comment (maxandersen) [reply devtools/aesh/src/main/java/io/quarkus/cli/QuarkusCli.java:9]: ah neve rmind - this whole sub project is setup as non-quarkus....as command mode didn't exist then. something for future update.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:26]: this would be ignoring ./gradlew / gradle wrapper

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:55]: this would be ignoring ./mvnw / maven wrapper

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (maxandersen) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:26]: nm - wrapper is done inanother method.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:85]: why is it that you can't just use the same way of running the wrapper and the path executable ?

**Comment by maxandersen** [comment]
> To create a native image use: `mvn install -Dnative-image`

`mvn install -Pnative-image` is the correct one ;)

**Comment by maxandersen** [comment]
on first review i agree with alexey that this PR doesn't  hurt going in - it does some basic cleanup which is good. I'll @stalep and @ia3andy  figure out which pr goes first ;) 

That said I think there are some things to consider what the first quarkus cli we push should do. I'll add my list of questions/concerns here; and tomorrow i'll get it added to a doc/issue for quarkus cli:

- i think it would be good the cli is implemented as a command mode quarkus app (drink our own champagne)
- the command layout is  verbose and should be revised.
   i.e. `quarkus list-extensions` vs the more 'modern' verb based `quarkus extension list`   
- 
- should we stay using aesh or move to picocli+jline (I already talked with @stalep who is ok with either or :)

**Comment by maxandersen** [comment]
btw. native-image failed to build for me on OSX:

```
[quarkus-cli-999-SNAPSHOT:91985]     analysis:  15,391.21 ms,  2.29 GB
Error: Classes that should be initialized at run time got initialized during image building:
 org.jboss.logging.JDKLevel was unintentionally initialized at build time. To see why org.jboss.logging.JDKLevel got initialized use -H:+TraceClassInitialization
org.slf4j.helpers.NamedLoggerBase was unintentionally initialized at build time. To see why org.slf4j.helpers.NamedLoggerBase got initialized use -H:+TraceClassInitialization
org.slf4j.helpers.Util was unintentionally initialized at build time. To see why org.slf4j.helpers.Util got initialized use -H:+TraceClassInitialization
org.jboss.slf4j.JBossLoggerAdapter was unintentionally initialized at build time. To see why org.jboss.slf4j.JBossLoggerAdapter got initialized use -H:+TraceClassInitialization
org.slf4j.helpers.MarkerIgnoringBase was unintentionally initialized at build time. To see why org.slf4j.helpers.MarkerIgnoringBase got initialized use -H:+TraceClassInitialization
org.jboss.logmanager.AtomicArray was unintentionally initialized at build time. To see why org.jboss.logmanager.AtomicArray got initialized use -H:+TraceClassInitialization
```

File: devtools/aesh/pom.xml
Comment (stalep) [reply devtools/aesh/pom.xml:23]: ATM I'm directly calling the Gradle binary or wrapper for compile/dev-mode. I noticed the gradle-api, but it is not on mavencentral. I commented it out for this PR as I would like to play more with that api to see if it's possible to call Gradle native instead of starting a separate process.

File: devtools/aesh/src/main/java/io/quarkus/cli/QuarkusCli.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/QuarkusCli.java:9]: This was written before command mode so I never thought about it :) 
What are the benefits of using quarkus command mode here?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:85]: If we detect a maven install we use the maven-invocation api to kick off the build. That is what I'll try to do with the gradle-api as well, but since they are called differently now I had to distinct wrapper and executable.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/ExecuteUtil.java:55]: There is a check first to find out if there is a wrapper present or not.

File: devtools/aesh/reflectconfigs/quarkuscli.json
Comment (stalep) [reply devtools/aesh/reflectconfigs/quarkuscli.json:6]: I agree :)

**Comment by stalep** [comment]
> > To create a native image use: `mvn install -Dnative-image`
> 
> `mvn install -Pnative-image` is the correct one ;)

Doh! - thanks

**Comment by stalep** [comment]
> on first review i agree with alexey that this PR doesn't hurt going in - it does some basic cleanup which is good. I'll @stalep and @ia3andy figure out which pr goes first ;)

@ia3andy will get his in first, I'll rebase this on top of that when it's in.

> That said I think there are some things to consider what the first quarkus cli we push should do. I'll add my list of questions/concerns here; and tomorrow i'll get it added to a doc/issue for quarkus cli:
> 
> * i think it would be good the cli is implemented as a command mode quarkus app (drink our own champagne)

We can do, I do not know enough about command mode to say anything against it. ATM the CLI can do anything Maven/Gradle can do, but with command mode I guess we can do even more? 

> * the command layout is  verbose and should be revised.
>   i.e. `quarkus list-extensions` vs the more 'modern' verb based `quarkus extension list`

That won't be an issue, æsh has support for as many levels of child/group commands as you want. Originally I just wanted to use as similar names to Maven/Gradle as possible, but I agree with your point.

> * should we stay using aesh or move to picocli+jline (I already talked with @stalep who is ok with either or :)

Yup, I'm fine with either way. I do not know picocli though and won't probably spend that much time on it (they're usually implementing whatever æsh has eventually anyway :)

**Comment by maxandersen** [comment]
> > * i think it would be good the cli is implemented as a command mode quarkus app (drink our own champagne)
> 
> We can do, I do not know enough about command mode to say anything against it. ATM the CLI can do anything Maven/Gradle can do, but with command mode I guess we can do even more?

The main thing is that the whole compile for native is there and we can use the ecoystem of extensions for whatever we need and it becomes a nice drive/testcase for quarkus building natively on more than just native linux :)

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java:17]: “build” seems more appropriate ?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java:44]: Not critical now but I would say package rather than install would be done here.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java:38]: create as we use that everywhere else.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:34]: Default is true to match elsewhere ?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:141]: There are no shared base class for these exceptions ?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/InstallExtensionCommand.java
Comment (maxandersen) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/InstallExtensionCommand.java:29]: We won’t actually be doing install, just adding extension. 

add, maybe ?

**Review by maxandersen — CHANGES_REQUESTED** [review]
A few suggestions / clarifications, but otherwise as stated before. Better than what is there now this let’s get it in :)

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (aloubyansky) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:23]: Only in Maven.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (aloubyansky) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java:38]: Everywhere else is Maven?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (aloubyansky) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java:41]: Does case matter, btw?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:141]: No, but that's a good idea, I'll implement that in Æsh asap

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java:41]: If the name is not set, Æsh will use the field name (including cases) as the name. I personally try to default option names to lower case, but I guess that is mostly personal preference.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java]: I changed it to dev, TBH I do not have much preferences for names.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java]: I renamed it to ```init``` as it was mentioned in a thread recently and I thought it was good. I do not care that much though. I changed it to ```create``` now.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java]: Fixed

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CompileProjectCommand.java]: Yup

**Comment by stalep** [comment]
I pushed some changes now that should reflect the latest comments. I see that github complains about conflicting files, but I did not get any conflicts locally when I rebased to master...

File: devtools/aesh/pom.xml
Comment (ia3andy) [reply devtools/aesh/pom.xml:23]: @stalep with this, is it possible to use the gradle api?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/BuildToolConverter.java
Comment (ia3andy) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/BuildToolConverter.java:14]: Maybe we should throw an exception here (instead of returning null)?

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (ia3andy) [inline devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:113]: Maybe should make this static method part of `BuildTool` (it's my mistake)?

**Review by ia3andy — APPROVED** [review]
Cool :-)

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/BuildToolConverter.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/BuildToolConverter.java:14]: Yes, probably wise. I didn't care as much as the null reference is handled by Æsh, but yeah. I'll change it.

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/DevModeCommand.java:113]: Yes, that's probably useful. I can just change that in this commit. Give me a few min (lunch now).

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java
Comment (maxandersen) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/CreateProjectCommand.java:41]: lower case over mixed case - such a pain in mvn ;)

**Comment by stalep** [comment]
Updated to resolve the latest comments from Andy

File: devtools/aesh/src/main/java/io/quarkus/cli/commands/InstallExtensionCommand.java
Comment (stalep) [reply devtools/aesh/src/main/java/io/quarkus/cli/commands/InstallExtensionCommand.java:29]: Yeah, as I mentioned earlier as well I used some names that were brought up in a thread I read. I thought install worked well, but we can use add too.

**Comment by stalep** [comment]
> A few suggestions / clarifications, but otherwise as stated before. Better than what is there now this let’s get it in :)

Max, I can't merge as you requested changes. Not sure if it was intentional or not? :)

**Review by maxandersen — APPROVED** [review]
looks good to go in.

**Pull request merged**
